content/blog/unsafe-string-interning-in-go.md

```diff
@@ -0,0 +1,345 @@
+pointer field. It is highly likely that a small number of zone identifiers will
+be used repeatedly throughout a program, so string interning seems like a great
+fit for this problem.
+
+The result of this work is the package
+[`go4.org/intern`](https://go4.org/intern) which uses some pretty neat `unsafe`
+tricks to implement efficient string interning using weak references and Go
+finalizers. We'll start by showing off the safe implementation and gradually
+introduce the concepts needed to understand the unsafe one as well.
+
+All credit for the ideas and implementation of `go4.org/intern` goes to [Dave
+Anderson](https://twitter.com/dave_universetf), [Josh Bleecher
+Snyder](https://twitter.com/commaok), and [Brad
+Fitzpatrick](https://twitter.com/bradfitz). I simply watched the discussions
+unfold on the issue tracker and thought that the code and the ideas behind it
+were too interesting to not write up and share!
+
+## Safe string interning in Go
+
+The basic idea of string interning is to retain a global pool of strings which
+can be created or fetched on the fly, reducing the number of repeated
+allocations for values which are identical. In its most basic form, you could
+implement `go4.org/intern` (with a caveat!) by writing the following:
+
+```go
+package intern
+
+// A Value pointer is the handle to an underlying comparable value.
+// See func Get for how Value pointers may be used.
+type Value struct {
+	// Enforce that Values cannot be used directly for equality comparisons.
+	// We'll explain why a bit later.
+	_      [0]func()
+	cmpVal interface{}
+}
+
+// Get returns the comparable value passed to the Get func
+// that returned v.
+func (v *Value) Get() interface{} { return v.cmpVal }
+
+// Our pool of interned values and a lock to serialize access.
+var (
+	mu      sync.Mutex
+	val = map[interface{}]*Value{}
+)
+
+// Get returns a pointer representing the comparable value cmpVal.
+//
+// The returned pointer will be the same for Get(v) and Get(v2)
+// if and only if v == v2, and can be used as a map key.
+func Get(cmpVal interface{}) *Value {
+	mu.Lock()
+	defer mu.Unlock()
+
+	v := val[cmpVal]
+	if v != nil {
+		// Value is already interned in the pool.
+		return v
+	}
+
+	// Value must be created now and then interned.
+	v = &Value{cmpVal: cmpVal}
+	val[cmpVal] = v
+	return v
+}
+```
+
+Usage of the package is straightforward at this point:
+
+```go
+func TestBasics(t *testing.T) {
+	foo := intern.Get("foo")
+	foo2 := intern.Get("foo")
+
+	if foo.Get() != foo2.Get() {
+		t.Error("foo values differ")
+	}
+	if foo.Get() != "foo" {
+		t.Error("foo.Get not foo")
+	}
+	if foo != foo2 {
+		t.Error("foo pointers differ")
+	}
+}
+```
+
+The global pool in the library stored the string `"foo"` on the first call to
+`intern.Get("foo")` and the same `*intern.Value` was returned and stored in both
+`foo` and `foo2`.
+
+However there is a caveat to this approach! If we keep adding more and more
```

Comment:
That'll always be true: If we keep adding more and more, it'll grow in size. The key point here is that it *never shrinks*. Even if you're no longer using a value, it remains in the pool.